src/learning/pca.rs

```diff
@@ -0,0 +1,203 @@
+    pub fn components(&self) -> LearningResult<&Matrix<f64>>  {
+        match self.components {
+            None => Err(Error::new_untrained()),
+            Some(ref rot) => { Ok(rot) }
+        }
+    }
+}
+
+/// The default PCA.
+///
+/// Parameters:
+///
+/// - `n` = `None` (keep all components)
+/// - `center` = `true`
+///
+/// # Examples
+///
+/// ```
+/// use rusty_machine::learning::pca::PCA;
+///
+/// let model = PCA::default();
+/// ```
+impl Default for PCA {
+    fn default() -> Self {
+        PCA {
+            // because number of columns is unknown,
+            // return all components by default
+            n: None,
+            center: true,
+
+            n_features: None,
+            centers: None,
+            components: None
+        }
+    }
+}
+
+/// Train the model and predict the model output from new data.
+impl UnSupModel<Matrix<f64>, Matrix<f64>> for PCA {
+
+    fn predict(&self, inputs: &Matrix<f64>) -> LearningResult<Matrix<f64>>  {
+
+        match self.components {
+            None => Err(Error::new_untrained()),
+            Some(ref comp) => {
+
+                match self.n_features {
+                    // this can't happen
+                    None => {
+                        return Err(Error::new_untrained());
+                    },
+                    Some(f) => {
+                        if f != inputs.cols() {
+                            return Err(Error::new(ErrorKind::InvalidData,
+                                       "Input data does not have the same dimensions as training data"));
+                        }
+                    }
+                };
+
+                if self.center == true {
+                    match self.centers {
+                        // this can't happen
+                        None => return Err(Error::new_untrained()),
+                        Some(ref centers) => {
+                            let data = centering(inputs, &centers);
+                            Ok(data * comp)
+                        }
+                    }
+                } else {
+                    Ok(inputs * comp)
+                }
+            }
+        }
+    }
+
+    fn train(&mut self, inputs: &Matrix<f64>) -> LearningResult<()> {
+        let data = if self.center == true {
+            let centers = inputs.mean(Axes::Row);
+            let m = centering(inputs, &centers);
+            self.centers = Some(centers);
+            m
+        } else {
+            inputs.clone()
+        };
+        let (_, _, v) = data.svd().unwrap();
+
+        self.components = match self.n {
+            Some(c) => {
+                let slicer: Vec<usize> = (0..c).collect();
+                Some(v.select_cols(&slicer))
+            },
+            None => Some(v)
+        };
+        self.n_features = Some(inputs.cols());
+        Ok(())
+    }
+}
+
+/// Subtract center Vector from each rows
+fn centering(inputs: &Matrix<f64>, centers: &Vector<f64>) -> Matrix<f64> {
```

Comment:
This is minor but I think we should make this function unsafe. This way the emphasis is on whoever is using it to make sure it works properly. In your case it looks fine but if someone else makes changes later it would be a welcome reminder.